### PR TITLE
AWB with Neural Networks

### PR DESCRIPTION
**Pull request by davidplowman**
Initial implementation of AWB using neural networks. This is an alternative to the old Bayesian method.

### CODE REVIEW

File: src/ipa/rpi/controller/rpi/awb.cpp [hunk 1]
```diff
@@ -476,11 +398,7 @@ void Awb::prepareStats()
 	 * LSC has already been applied to the stats in this pipeline, so stop
 	 * any LSC compensation.  We also ignore config_.fast in this version.
 	 */
-	const double biasCtR = config_.bayes ? config_.ctR.eval(config_.biasCT) : 0;
-	const double biasCtB = config_.bayes ? config_.ctB.eval(config_.biasCT) : 0;
-	generateStats(zones_, statistics_, config_.minPixels,
-		      config_.minG, getGlobalMetadata(),
-		      config_.biasProportion, biasCtR, biasCtB);
+	generateStats(zones_, statistics_, 0.0, 0.0, getGlobalMetadata(), 0.0, 0.0, 0.0);
```
Comment (davidplowman) [inline src/ipa/rpi/controller/rpi/awb.cpp:401]: AwbBayes seems to have its own prepareStats, is that right? So is this the AwbNN version? Might be better to leave it pure virtual in the base class and define it in awb_nn.cpp.

File: src/ipa/rpi/controller/rpi/awb_nn.cpp [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+	AwbNNConfig() {}
+	int read(const libcamera::YamlObject &params, AwbConfig &config);
+
+	/* An empty model will check default locations for model.tflite */
+	std::string model;
+	int numOutputs;
+	float minTemp;
+	float maxTemp;
+
+	bool enable_nn;
```
Comment (davidplowman) [inline src/ipa/rpi/controller/rpi/awb_nn.cpp:46]: libcamera will want `enableNn` instead of `enable_nn`.

File: src/ipa/rpi/controller/rpi/awb_nn.cpp [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+{
+	LOG(RPiAwb, Debug) << "Loading AWB model " << nnConfig_.model;
+	std::string modelPath = "/ipa/rpi/pisp/awb_model.tflite";
+	if (nnConfig_.model.empty()) {
+		std::string root = utils::libcameraSourcePath();
+		if (!root.empty()) {
+			modelPath = root + modelPath;
+		} else {
+			modelPath = LIBCAMERA_DATA_DIR + modelPath;
+		}
```
Comment (davidplowman) [inline src/ipa/rpi/controller/rpi/awb_nn.cpp:161]: Might be worth printing out the model path in an Info statement. People waste so much time thinking they're loading something that they aren't (myself included)!

Comment (davidplowman) [reply src/ipa/rpi/controller/rpi/awb_nn.cpp:161]: Ah I see we do it once it's loaded. Well, myabe a Debug statement then, so that folks can double-check if it's not being found.

File: src/ipa/rpi/controller/rpi/awb_nn.cpp [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+		loadModel();
+		if (!nnConfig_.enable_nn) {
+			LOG(RPiAwb, Warning) << "Neural Network AWB failed to load - switch to Grey method";
+			config_.greyWorld = false;
+			config_.sensitivityR = config_.sensitivityB = 1.0;
+		}
+	}
+}
+
+void AwbNN::transverseSearch(double t, double &r, double &b)
```
Comment (davidplowman) [inline src/ipa/rpi/controller/rpi/awb_nn.cpp:249]: I sometimes wonder if we should do a very short search along the curve too, but not sure. TBD, I guess.

File: src/ipa/rpi/controller/rpi/awb_nn.cpp [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+								   points[bestPoint],
+								   points[bestPoint + 1]);
+	double rBest = rbBest.x(), bBest = rbBest.y();
+
+	r = rBest, b = bBest;
+}
+
+static double getElementPadded(const std::vector<double> &array, int i, int j)
+{
+	int i_padded = std::clamp(i, 0, 31);
```
Comment (davidplowman) [inline src/ipa/rpi/controller/rpi/awb_nn.cpp:299]: libcamera will want snakecase `iPadded`.

File: src/ipa/rpi/controller/rpi/awb_nn.cpp [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+	return array[i_padded * 32 + j_padded];
+}
+
+static std::vector<AwbNN::RGB> filter(const std::vector<AwbNN::RGB> &rgb)
+{
+	/*
+	 * Decreases the difference between the colour channels when it is higher
+	 * than nearby pixels.
+	 */
+	const int total_pixels = 32 * 32;
```
Comment (davidplowman) [inline src/ipa/rpi/controller/rpi/awb_nn.cpp:310]: Same throughout this function, `totalPixels` etc.

File: src/ipa/rpi/controller/rpi/awb_nn.cpp [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+
+AwbNN::RGB AwbNN::processZone(AwbNN::RGB zone, float lux, float red_gain, float blue_gain)
+{
+	/*
+	 * Renders the pixel at 5000K temperature then takes the log of the
+	 * r/g and b/g ratios. The log is then scaled to 0-1 and clamped.
+	 * These are then combined with the lux to get the input for the.
+	 * neural network.
+	 */
+	RGB zone_gains = zone;
```
Comment (davidplowman) [inline src/ipa/rpi/controller/rpi/awb_nn.cpp:382]: `zoneGains` etc.

File: src/ipa/rpi/controller/rpi/awb_nn.cpp [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
+	zone_log.R = std::clamp(zone_log.R, 0.0, 1.0);
+	zone_log.G = std::clamp(zone_log.G, 0.0, 1.0);
+	zone_log.B = std::clamp(zone_log.B, 0.0, 1.0);
+
+	return zone_log;
+}
+
+static float interpolateOutput(float *output_data, int num_outputs, float min_temp, float max_temp)
+{
+	float total_value = 0.0;
```
Comment (davidplowman) [inline src/ipa/rpi/controller/rpi/awb_nn.cpp:423]: `totalValue` etc.

File: src/ipa/rpi/controller/rpi/awb_nn.cpp [hunk 9]
```diff
@@ -0,0 +1,505 @@
...
+		total_value += temp * weight;
+		total_weight += weight;
+	}
+
+	return total_value / total_weight;
+}
+
+void AwbNN::awbNN()
+{
+	float *input_data = interpreter_->typed_input_tensor<float>(0);
```
Comment (davidplowman) [inline src/ipa/rpi/controller/rpi/awb_nn.cpp:439]: `inputData` etc.

File: src/ipa/rpi/controller/rpi/awb_nn.cpp [hunk 10]
```diff
@@ -0,0 +1,505 @@
...
+			LOG(RPiAwb, Error) << "CT curve not specified";
+			enable_nn = false;
+		}
+
+		if (!model.empty() && model.find(".tflite") == std::string::npos) {
+			LOG(RPiAwb, Error) << "Model must be a .tflite file";
+			enable_nn = false;
+		}
+
+		bool valid_ccm = true;
```
Comment (davidplowman) [inline src/ipa/rpi/controller/rpi/awb_nn.cpp:99]: `validCcm` etc.

File: src/ipa/rpi/controller/rpi/awb_nn.cpp [hunk 11]
```diff
@@ -0,0 +1,505 @@
...
+	interpreter_->AllocateTensors();
+	TfLiteTensor *input_tensor = interpreter_->input_tensor(0);
+	TfLiteTensor *output_tensor = interpreter_->output_tensor(0);
+	if (!input_tensor || !output_tensor) {
+		LOG(RPiAwb, Error) << "Model missing input or output tensor";
+		nnConfig_.enable_nn = false;
+		return;
+	}
+
+	const int expected_input_size = 32 * 32 * 3;
```
Comment (davidplowman) [inline src/ipa/rpi/controller/rpi/awb_nn.cpp:199]: libcamera will want `expectedInputSize`. Same for all the others...

File: src/ipa/rpi/controller/rpi/awb_nn.cpp [hunk 12]
```diff
@@ -0,0 +1,505 @@
...
+
+	zone_log.R = std::log(r_ratio);
+	zone_log.G = std::log(lux) / 10;
+	zone_log.B = std::log(b_ratio);
+
+	zone_log.R += 1.0;
+	zone_log.B += 1.0;
+
+	zone_log.R /= 2.0;
+	zone_log.B /= 2.0;
```
Comment (davidplowman) [inline src/ipa/rpi/controller/rpi/awb_nn.cpp:412]: Is there any particular rationale for these adjustment? Might be worth a comment.

File: src/ipa/rpi/controller/rpi/awb_nn.cpp [hunk 13]
```diff
@@ -0,0 +1,505 @@
...
+
+static float interpolateOutput(float *output_data, int num_outputs, float min_temp, float max_temp)
+{
+	float total_value = 0.0;
+	float total_weight = 0.0;
+
+	for (int i = 0; i < num_outputs; i++) {
+		float temp = min_temp + i * (max_temp - min_temp) / (num_outputs - 1);
+		float weight = output_data[i];
+		weight *= weight;
```
Comment (davidplowman) [inline src/ipa/rpi/controller/rpi/awb_nn.cpp:429]: Interesting. Makes all the weights +ve, among other things! I wonder a bit about putting a softmax on the end of the network instead, is that something we've tried?
